packages/react-router-dom/modules/__tests__/Focus-test.js

```diff
@@ -0,0 +1,439 @@
+import "jest";
+import React from "react";
+import ReactDOM from "react-dom";
+import { createMemoryHistory } from "history";
+import { MemoryRouter, Router, Switch, Route, Focus } from "react-router-dom";
+
+import renderStrict from "./utils/renderStrict";
+
+jest.useFakeTimers();
+
+describe("<Focus>", () => {
+  const node = document.createElement("div");
+
+  afterEach(() => {
+    ReactDOM.unmountComponentAtNode(node);
+  });
+
+  describe("mounting", () => {
+    it("focuses ref component when mounting", () => {
+      renderStrict(
+        <MemoryRouter>
+          <Focus>
+            {ref => (
+              <div id="test" tabIndex={-1} ref={ref}>
+                Testing!
+              </div>
+            )}
+          </Focus>
+        </MemoryRouter>,
+        node
+      );
+
+      jest.runAllTimers();
+
+      const ref = node.querySelector("#test");
+      const focused = document.activeElement;
+      expect(focused).toBe(ref);
+    });
+
+    it("warns if ref isn't attached to an element (body focused)", () => {
+      jest.spyOn(console, "warn").mockImplementation(() => {});
+
+      renderStrict(
+        <MemoryRouter>
+          <Focus>{ref => null}</Focus>
+        </MemoryRouter>,
+        node
+      );
+
+      jest.runAllTimers();
```

Comment:
Whoa ... what magic is *this*?! Could this help me get rid of [my janky `waitForRedirect`](https://github.com/ReactTraining/react-router/blob/5ccb241cd59eae7a15de31c7babff6619378be39/packages/react-router/modules/__tests__/Switch-test.js#L65-L68) function?? 😅

Reply:
[Timer mocks](https://jestjs.io/docs/en/timer-mocks.html) are great.